### ISSUE TITLE
Qwen3.6-27B-FP8 fails to load and produces corrupted output on llm-scaler-vllm b8.3.1 (XPU, TP=2)

### ISSUE THREAD

**Issue by bongmiin** [intel/llm-scaler#474]
## Summary

On `intel/llm-scaler-vllm:0.14.0-b8.3.1`, the official `Qwen/Qwen3.6-27B-FP8` checkpoint fails to load on XPU. After working around the load-time failures, the server boots but generation is corrupted.

I traced this to four separate issues. Issues 1–3 are load/parameter-registration problems that can be worked around with local source patches. Issue 4 appears to be related to the dense MLP FP8 ESIMD fast path.

With the local patches for issues 1–3 and the following runtime flags:

```bash
DISABLE_ESIMD_FUSED_INPUT=1
DISABLE_ESIMD_DENSE=1
```

the model serves correctly on 2x Intel Arc Pro B70 with tensor parallelism 2.

## Environment

* Image: `intel/llm-scaler-vllm:0.14.0-b8.3.1`
* vLLM build: `0.14.1.dev0+gb17039bcc.d20260605`
* Model: `Qwen/Qwen3.6-27B-FP8`
* Quantization: FP8 block-wise, `weight_block_size=[128,128]`
* Model structure: split `gate_proj` / `up_proj`, `in_proj_qkv` / `in_proj_z`
* Hardware: 2x Intel Arc Pro B70, Battlemage
* Parallelism: `tensor-parallel-size=2`
* dtype: `float16`
* max model length: `40000`
* mode: `enforce-eager`

---

## Issue 1: load fails with `MergedColumnParallelLinear object has no attribute weight`

During FP8 `process_weights_after_loading`, loading fails with:

```text
'MergedColumnParallelLinear' object has no attribute 'weight'
```

Before the crash, many parameters are skipped:

```text
Parameter layers.N.mlp.gate_proj.weight not found in params_dict, skip loading
Parameter layers.N.linear_attn.in_proj_qkv.weight not found in params_dict, skip loading
```

In some runs, corrupted parameter names also appear:

```text
gate_gate_up_proj
qkqkv_proj
```

### Root cause

In `vllm/patches/vllm_for_multi_arc.patch`, the added `qwen3_5.py` and `qwen3_5_mtp.py` `load_weights()` implementation mutates `name` in place inside the `stacked_params_mapping` loop:

```python
name = name.replace(weight_name, param_name)

if name.endswith(".bias") and name not in params_dict:
    continue
if is_pp_missing_parameter(name, self):
    continue
if name not in params_dict:
    continue

param = params_dict[name]
```

If a mapped name is missing, the loop continues with the already-mutated `name`. A later mapping entry can then apply `.replace()` again to an already mapped name.

Because names such as `up_proj` are substrings of `gate_up_proj`, and `q_proj` / `k_proj` are substrings of `qkv_proj`, this can produce invalid double-mapped names:

```text
gate_proj -> gate_up_proj -> gate_gate_up_proj
q_proj/k_proj -> qkv_proj -> qkqkv_proj
```

The expert branch below this code already uses the safer pattern: create a local mapped name, validate it, and only commit the mapped name on success.

### Local fix

Use a local candidate name and only assign it back to `name` after a successful lookup:

```python
name_mapped = name.replace(weight_name, param_name)

if name_mapped.endswith(".bias") and name_mapped not in params_dict:
    continue
if is_pp_missing_parameter(name_mapped, self):
    continue
if name_mapped not in params_dict:
    continue

name = name_mapped
param = params_dict[name]
```

The same fix is needed in `qwen3_5_mtp.py`.

After this change, the corrupted `gate_gate_up_proj` / `qkqkv_proj` warnings disappear.

---

## Issue 2: fused FP8 modules still expose only bias after Issue 1 fix

After fixing the loader name mutation, the fused modules still expose only `bias` in `named_parameters()`:

```text
layers.42.linear_attn.in_proj_qkvz  params=['bias']
layers.42.mlp.gate_up_proj          params=['bias']
layers.42.mlp.down_proj             params=['bias']
```

This means serialized FP8 weights/scales are not registered on these fused modules.

### Root cause

`XPUFp8LinearMethod.create_weights()` does not appear to register the serialized checkpoint FP8 parameters for this path. For serialized checkpoint FP8, delegating to the parent `create_weights()` path fixes the missing parameters.

### Local fix

```python
if self.quant_config.is_checkpoint_fp8_serialized:
    return super().create_weights(
        layer=layer,
        input_size_per_partition=input_size_per_partition,
        output_partition_sizes=output_partition_sizes,
        input_size=input_size,
        output_size=output_size,
        params_dtype=params_dtype,
        **extra_weight_attrs,
    )
```

After this change, the modules expose the expected parameters, including:

```text
weight
weight_scale_inv
input_scale
```

and all 66 shards load.

---

## Issue 3: `MergedColumnParallelLinear object has no attribute weight_scale`

After fixing Issues 1 and 2, loading fails again with:

```text
'MergedColumnParallelLinear' object has no attribute 'weight_scale'
```

### Root cause

For block-wise FP8, the parent `process_weights_after_loading()` path keeps the scale parameter as `weight_scale_inv`. However, the XPU fused/GDN paths and the added Qwen loader path access `layer.weight_scale`.

### Local workaround

After `super().process_weights_after_loading(layer)` in the block-quant branch, alias `weight_scale_inv` to `weight_scale` when needed:

```python
if hasattr(layer, "weight_scale_inv") and not hasattr(layer, "weight_scale"):
    layer.register_parameter(
        "weight_scale",
        torch.nn.Parameter(layer.weight_scale_inv.data, requires_grad=False),
    )
```

This allows the model to finish loading.

---

## Issue 4: server boots, but generation is corrupted

After fixing Issues 1–3, the server boots and responds, but generation is corrupted.

Examples:

```text
Prompt: What is 2+2?
Output: 2+ 2 5

Prompt: Explain quantum ...
Output: markdown/LaTeX-like garbage
```

The reasoning and tool parsers appear to be applied correctly:

```bash
--reasoning-parser qwen3
--tool-call-parser qwen3_coder
```

so this does not look like a parser issue.

### Debugging notes

I checked the GDN path using `isfinite` / sum probes before and after `gdn_attention`; there was no immediate NaN.

I also compared layer-0 `in_proj_qkvz` fast path vs naive dequant, and the result was close.

However, setting:

```bash
DISABLE_ESIMD_DENSE=1
```

fixes generation.

Examples after disabling the dense ESIMD path:

```text
Prompt: What is 2+2?
Output: 4

Prompt: What is the capital of France?
Output: Paris

Prompt: Explain quantum ...
Output: coherent explanation
```

Therefore, the dense MLP FP8 ESIMD fast path, `_dense_mlp_fp8`, appears to corrupt logits for this model/shape.

Additionally, I also needed:

```bash
DISABLE_ESIMD_FUSED_INPUT=1
```

because `in_proj_ba` is excluded from quantization, so the ESIMD projection path tries to access `in_proj_ba.weight_scale`, which does not exist.

---

## Net result

With:

1. the Qwen loader name-mutation fix,
2. serialized FP8 parameter registration for XPU fused modules,
3. `weight_scale_inv` / `weight_scale` compatibility handling,
4. `DISABLE_ESIMD_FUSED_INPUT=1`,
5. `DISABLE_ESIMD_DENSE=1`,

`Qwen/Qwen3.6-27B-FP8` serves correctly on:

```text
2x Intel Arc Pro B70
tensor-parallel-size=2
dtype=float16
max-model-len=40000
enforce-eager
```

## PR offer

I am happy to open a PR for Issue 1, the Qwen loader name-mutation bug, against `vllm_for_multi_arc.patch`.

Issues 2–4 likely need maintainer guidance on the intended XPU FP8 parameter-registration and ESIMD fast-path design.

**Comment by parthvzala** [comment]
Are you able to use it for coding, or it still not enough?

**Comment by nstiurca** [comment]
Regarding Issue 3,

> After super().process_weights_after_loading(layer) in the block-quant branch, alias weight_scale_inv to weight_scale when needed:

```python
if hasattr(layer, "weight_scale_inv") and not hasattr(layer, "weight_scale"):
    layer.register_parameter(
        "weight_scale",
        torch.nn.Parameter(layer.weight_scale_inv.data, requires_grad=False),
    )
```
> This allows the model to finish loading.

Erm, this can't be right, can it? `_inv` as in inverse, as in 1/weight_scale, no?? Probably why you get garbled outputs later. Disabling the ESIMD stuff in step 4 apparently results in not using this mis-aliased parameter. Or am I taking crazy pills??

**Comment by bongmiin** [comment]
@parthvzala 
Qwen3.6 itself seems to be a decent model for coding. That impression is based on my experience running it on an NVIDIA A100 80GB PCIe.

The main issue is the serving setup: with TP=2, the model barely fits because of its size, and inference is very slow.

For practical use, especially for coding tasks, I think it would likely require TP=4 or higher.

**Comment by gc-fu** [comment]
Hi, we currently do not support block quantization models such as `Qwen/Qwen3.6-27B-FP8`.  We will support this type of quantization in later releases.

**Comment by bongmiin** [comment]
@nstiurca 
Thanks! Yeah, the name trips everyone up, but it's actually fine here. In DeepSeek-style block FP8 (which Qwen3.6 uses), `weight_scale_inv` is already the multiply-by factor for dequant — `W = W_fp8 * weight_scale_inv`, not `1/weight_scale`. So copying it straight into `weight_scale` is correct; inverting it would be the bug.

The garbled output isn't from the alias though — the non-ESIMD path uses the same values and works fine. It's `_dense_mlp_fp8` mishandling the block-wise scale, which is why `DISABLE_ESIMD_DENSE=1` fixes it.

**Comment by bongmiin** [comment]
Thanks for the clarification.

That explains the behavior I was seeing with `Qwen/Qwen3.6-27B-FP8`. I’ll close this issue for now since block-wise FP8 quantization is not officially supported yet.

I’ll try again once support for this quantization type is available in a later release.
